src/app/store/observeMiddleware.ts

```diff
@@ -0,0 +1,76 @@
+import { deepEqual } from 'fast-equals';
+import { Dispatch, isAction, Middleware, MiddlewareAPI } from 'redux';
+import { createAction, createCustomAction, isActionOf } from 'typesafe-actions';
+import { RootState } from './types';
+
+interface StoreObserver<T> {
+  id: string;
+  /**
+   * Function to create "something" which will be used to determine if the side effect should run.
+   * deepEqual is used for the comparison, so use primitives, shallow objects or arrays to keep
+   * performance reasonable.
+   */
+  getObserved: (rootState: RootState) => T;
+  /**
+   * Runs the side effect providing both the previous and current version of the derrived state.
+   */
+  sideEffect: (states: { previous: T; current: T }) => void;
+}
+
+// Need to user a higher order function to get the correct typings and inference, it will now correctly
+// type the value of T based on what the getObserved returns
+export const observe = <T>(storeObserver: StoreObserver<T>) =>
+  createCustomAction('observer/OBSERVE', (storeObserver: StoreObserver<T>) => ({ storeObserver }))(
+    storeObserver,
+  );
+export const unobserve = createAction('observer/UNOBSERVE')<string>();
+export const clearObservers = createAction('observer/CLEAR_OBSERVERS')();
```

Comment:
Are we going to run into similar issues with the action creators being located here 🤔? 

The only dependency I am utilising from the repo is the root state, which is just a type so I think it should be easily stripped at compile time and not get in the way of chunk splitting. 

The question then remains on best location for the middleware file.

Reply:
Shouldn't be a problem, as the middleware doesn't depend on the store, the store depends on the middleware. RootState was moved to a types file specifically to break that dependency.

Reply:
I tried initially but it didn't seem to work with the custom action. 

I'll have another go when I get back to the PR and report back the type issues I see (if they still exist)